File: Assignments/AlgorithmLib/BetterLinearSearch.cs
/* CSE 381 - BetterLinearSerach
*  (c) BYU-Idaho - It is an honor code violation to post this
*  file completed in a public file sharing site.  F6.
*
*  Instructions: Refer to W01 Prove: Assignment in Canvas for detailed instructions.
*/

using Microsoft.VisualBasic;

namespace AlgorithmLib;
List<int> dataSet = new();
public static class BetterLinearSearch
{

// add checks for empty list and 1 item lists
    if(dataSet.Count == 0) 
    {
        return -1;
    }

// loop to find value 
    /* Search for an item in a list.  Ignore duplicates by exiting
    *  as soon as the first match is found.
    *
    *  Inputs:
    *     data - list to search
    *     target - value to search for
    *  Outputs:
    *     Index where target was found
    *
    *  Note: Return -1 if target not found
    */
    public static int Search<T>(List<T> data, T target) where T : IComparable<T>
    {
        return 0;
    }
}

// turn in screenshot of code and of passing tests 

File: Assignments-Rust/src/better_linear_search.rs
/* CSE 381 - BetterLinearSerach
*  (c) BYU-Idaho - It is an honor code violation to post this
*  file completed in a public file sharing site.  F6.
*
*  Instructions: Refer to W01 Prove: Assignment in Canvas for detailed instructions.
*/

// NOTE: These statements will supress warnings in your code.  It is recommended
// that you comment these out when you start on this assignment
#![allow(unused_imports)]
#![allow(unused_variables)]
#![allow(dead_code)]

// NOTE: This warning should be supressed since you are re-implementing a 
// common find function in Rust.
#[allow(clippy::manual_find)]

/* Search for an item in a list.  Ignore duplicates by exiting
*  as soon as the first match is found.
*
*  Inputs:
*     data - list to search containing data that implements Eq (equality)
*     target - value to search for
*  Outputs:
*     An Option containing the index where the target was found: Some(index)
*
*  Errors: If the target is not found, then return the error Option: None
*/
pub fn search<T: Eq>(data : &[T], target : &T) -> Option<usize> {
    todo!()
} 

/* Do not modify these tests.
*  To run the tests, use the cargo tool from the command line:
*     Run All Tests: cargo test better_linear_search
*     Run One Test:  cargo test better_linear_search::tests::<test function name>
*/
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test1_match_number() {
        let data = [1, 2, 3, 4, 5, 6];
        let target = 4;
        assert_eq!(search(&data, &target), Some(3));
    }

    #[test]
    fn test2_no_match_number() {
        let data = [1, 2, 3, 4, 5, 6];
        let target = -1;
        assert_eq!(search(&data, &target), None);
    }

    #[test]
    fn test3_match_string() {
        let data = ["cat", "dog", "pig", "chicken", "duck"];
        let target = "pig";
        assert_eq!(search(&data, &target), Some(2));
    }

    #[test]
    fn test4_no_match_string() {
        let data = ["cat", "dog", "pig", "chicken", "duck"];
        let target = "eagle";
        assert_eq!(search(&data, &target), None);
    }
	
	#[test]
    fn test5_empty() {
        let data = [];
        let target = 42;
        assert_eq!(search(&data, &target), None);
    }
	
	#[test]
    fn test6_match_first() {
        let data = [1, 2, 3, 4, 5, 6];
        let target = 1;
        assert_eq!(search(&data, &target), Some(0));
    }
	
	#[test]
    fn test7_match_last() {
        let data = [1, 2, 3, 4, 5, 6];
        let target = 6;
        assert_eq!(search(&data, &target), Some(5));
    }
	
}